### ISSUE TITLE
Optimize retrieval of Status facet counts in Case Law Search

### ISSUE THREAD

**Issue by albertisfu** [freelawproject/courtlistener#4539]
Currently, for every search request in Case Law, we also query the facet counts for the status checkboxes.

![Screenshot 2024-10-07 at 3 35 48 p m](https://github.com/user-attachments/assets/888147d5-7c1b-4159-aa3b-f254283b9ee2)

We could optimize this by:
- Adding a cache for these facets. Similar to how we micro-cache results for each query, we can cache these counts, likely using the same cache duration.
- We use an aggregation on the `status` field to get the counts. It may be possible to combine this with a cardinality aggregation so the counts are approximate, making the query faster. Some testing with the production index will be required to assess the new query before implementing it.

**Comment by mlissner** [comment]
> likely using the same cache duration.

Can we use the same cache key too?

> the counts are approximate

With the result count, we were able to say "About XXX results" but we won't be able to do that here since saying "About" would take too much room. 

I think this might be time to make (or find, if it already exists) a fun little helper library that gives us approximations:

| Input | Example In | Output | Example output |
|--------|--------|---|---|
| n < 1,000 | 201 | The number itself | 201 |
| 1,000 ≤ n ≤ 999,999  | 5,899 | The number of thousands, rounded down, k | 5k |
| 1,000,000 ≤ n ≤ 999,999,999 | 24,233,000 | The number of millions, a period, the number of hundreds of thousands, rounded down, then M | 24.2M |

I think that should cover the bases and would nicely imply that the count is approximate.

**Comment by albertisfu** [comment]
> Can we use the same cache key too?

Yes, we can use the same cache key. All the Opinions queries also return facet counts, so it's fine to store them together.

> I think this might be time to make a fun little helper library that gives us approximations:

Great, yeah this should work nicely.

**Comment by mlissner** [comment]
Hi, this is Claude (AI assistant) working with @mlissner. We've analyzed this issue and developed an implementation plan. Documenting it here for reference.

## Implementation Plan

### Phase 1: Cardinality Aggregation + Caching for Facet Counts

**Step 1: Add cardinality aggregation to facet queries**

Modify `get_facet_dict_for_search_query()` and `get_only_status_facets()` in `cl/lib/elasticsearch_utils.py` to follow the existing cardinality pattern used by `build_cardinality_count()` and `simplify_estimated_count()`:

- Keep the `terms` aggregation on `status.raw` (needed for per-status bucket counts)
- Add cardinality sub-aggregation using `precision_threshold=settings.ELASTICSEARCH_CARDINALITY_PRECISION`
- Apply `simplify_estimated_count()` to bucket doc_counts in `make_es_stats_variable()` when counts exceed the threshold

**Step 2: Cache facet counts in the micro-cache**

Add a `facet_counts` field to `CachedESSearchResults` in `cl/lib/search_utils.py` to store the `{status: count}` mapping.

**Step 3: Add facet caching logic to `do_es_search()`**

- Build a facet-specific cache key from cleaned params **excluding** `stat_*` fields (so toggling status checkboxes reuses cached facet counts)
- Before calling `get_facet_dict_for_search_query()`, check the facet cache
- After computing facets, cache the dict with `SEARCH_RESULTS_MICRO_CACHE` duration

**Step 4: Add helper to reconstruct facet fields from cached dict**

Add `make_es_stats_variable_from_cache(search_form, facet_values)` in `cl/lib/elasticsearch_utils.py`.

### Phase 2: Approximate Count Display

**Step 5: Create approximate number helper**

Add `approximate_count(n: int) -> str` in `cl/lib/string_utils.py` (general-purpose, not search-specific):

| Input Range | Example In | Output |
|-------------|-----------|--------|
| n < 1,000 | 201 | "201" |
| 1,000 ≤ n < 1,000,000 | 5,899 | "5k" |
| 1,000,000+ | 24,233,000 | "24.2M" |

**Step 6: Create template filter**

Register an `approx_count` filter in `cl/lib/templatetags/extras.py` that calls the helper.

**Step 7: Update sidebar template**

Replace `{{ facet_field.count|intcomma }}` with `{{ facet_field.count|approx_count }}` in `cl/search/templates/includes/sidebar_filters.html`.

### Phase 3: Tests

- Unit tests for `approximate_count()` helper
- Facet caching tests (cache population, cache hit with different status params)
- Verify existing `test_facet_fields_counts` and `test_faceted_queries` still pass

### API Impact

None. The facet functions are frontend-only (called from `cl/search/views.py`). The REST API has its own separate code path and does not return status facet counts.
